lib/src/main/java/io/ably/lib/rest/Auth.java

```diff
@@ -128,16 +128,53 @@ public AuthOptions(String key) throws AblyException {
 		}
 
 		/**
-		 * Internal
+		 * Internal, when AuthOptions are provided, the values of each attribute are not merged,
+		 * but instead replace all corresponding values, even when null.
+		 *
+		 * should be used for:
+		 * - {@link Auth#requestToken(TokenParams, AuthOptions)} (Spec:RSA8e),
+		 * - {@link Auth#createTokenRequest(AuthOptions, TokenParams)} (Spec:RSA9h),
+		 * - {@link Auth#authorise(AuthOptions, TokenParams)} (Spec:RSA10j).
+		 *
+		 * <p>
+		 * Spec: RSA8e, RSA9h, RSA10j
+		 * </p>
+		 *
+		 * @param options if the object is null, the client library configured defaults are used.
+		 * @return superseded object
 		 */
-		public AuthOptions merge(AuthOptions defaults) {
-			if(authCallback == null) authCallback = defaults.authCallback;
-			if(authUrl == null) authUrl = defaults.authUrl;
-			if(key == null) key = defaults.key;
-			if(authHeaders == null) authHeaders = defaults.authHeaders;
-			if(authParams == null) authParams = defaults.authParams;
-			queryTime = queryTime & defaults.queryTime;
-			return this;
+		public AuthOptions replace(AuthOptions options) {
+			AuthOptions copy;
+			if (this instanceof ClientOptions) {
```

Comment:
Well, this is a bit more complicated... Question is: why do we extend ClientOptions from AuthOptions? And not including AuthOptions within ClientOptions? I.e. not using composition instead of inheritance? 

I've tried to find some examples of authorise in the docs but no luck so far. Trying to get a vision from the end user perspective. Technically user can pass ClientOptions to those methods as well. According to the docs let's say there is a reference like "token or tokenDetails attributes of ClientOptions" (http://docs.ably.io/rest/authentication/) which make me feels like I can definitely use it as AuthOptions. Also I might think that once I authorise using AuthOptions I might found my AuthOptions within AblyRest.options.

Just confusing a bit. If user can't pass ClientOptions to those methods then I think we should at least raise an exception telling that it's not a proper way of using it.

So all in all we just tried to cover different cases possible from the technical standpoint. Not doing checks for ClientOptions could be easier for sure.